device/src/io.rs

```diff
@@ -59,19 +68,32 @@ where
     T: esp32c3_hal::timer::Instance,
 {
     pub fn poll_read(&mut self) -> bool {
-        let read_buffer = &mut self.read_buffer;
+        let mut read_something = false;
         while let Ok(c) = self.io.read_byte() {
-            read_buffer.push(c);
+            self.ring_buffer.push(c);
+            read_something = true;
         }
-        !read_buffer.is_empty()
+        read_something
     }
 
     pub fn find_and_remove_magic_bytes(&mut self) -> bool
     where
         D: Direction,
     {
         self.poll_read();
-        frostsnap_comms::find_and_remove_magic_bytes::<D>(&mut self.read_buffer)
+        if self.ring_buffer.len() == 0 {
+            return false;
+        }
+
+        let (consume, progress, found) = frostsnap_comms::make_progress_on_magic_bytes::<D>(
+            &self.clone_buffer_to_vec()[..],
```

Comment:
The point of this PR is to remove allocations during the decoding on messages process but this does allocations. Why can't make progress on magic bytes just work with the ring buffer. Surely the API can be changed to work with whatever you pass in from the ring buffer.